File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        cur = head
        if not head:
            return head

        while cur:
            new = Node(cur.val)
            new.next = cur.next
            cur.next = new
            cur = new.next

        cur = head
        while cur:
            if cur.random:
                cur.next.random = cur.random.next
            cur = cur.next.next

        cur = head
        new_head = head.next
        new_cur = new_head
        while cur:
            cur.next = new_head.next
            if new_head.next:
                new_head.next = new_head.next.next
            cur = cur.next
            new_head = new_head.next

        return new_cur

        

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *next;
 *     struct Node *random;
 * };
 */

struct Node* createNode(int val) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->val = val;
    newNode->next = NULL;
    newNode->random = NULL;
    return newNode;
}

struct Node* copyRandomList(struct Node* head) {
    if (head == NULL) {
        return NULL;
    }

    struct Node* current = head;
    while (current != NULL) {
        struct Node* newNode = createNode(current->val);
        newNode->next = current->next;
        current->next = newNode;
        current = newNode->next;
    }

    current = head;
    while (current != NULL) {
        if (current->random != NULL) {
            current->next->random = current->random->next;
        }
        current = current->next->next;
    }

    current = head;
    struct Node* newHead = head->next;
    struct Node* newCurrent = newHead;
    while (current != NULL) {
        current->next = newCurrent->next;
        if (newCurrent->next != NULL) {
            newCurrent->next = newCurrent->next->next;
        }
        current = current->next;
        newCurrent = newCurrent->next;
    }

    return newHead;
}